### ISSUE TITLE
Lecture 13: Face recognition

### ISSUE THREAD

**Issue by kingheadcat** [KimDahye/deeplearning_study#13]
# Face recognition
---
# Face verification vs. face recognition
## Verification (Binary Classification)
- Input image, name/ID
- Output whether the input image is that of the claimed person
## Recognition (Multi Class Classification)
- Has a database of K persons
- Get an input image
- ID Output ID if the image is any of the K persons (or 
“not recognized”)

---
# Multi Class Classification vs Multi Label Classification 
![](2021-11-14-16-56-11.png)
- ref: [scikit-learn](https://scikit-learn.org/stable/modules/multiclass.html)
---
# One-shot learning: intro
## 하나의 example만으로 classifier를 학습하는 연구.  
few-shot learning으로 추상되며, 
보통 meta learning 방법론을 참조한다.

강의에서는 데이터들 간의 **유사도 함수를 학습**하는 형태의 방법론을 보였다.

---

# One-shot learning: Meta Learning 
**Meta refers to a level above.**
목표과제(labeled data)와는 간접적일 수 있는 목표함수(meta)를 학습하는 방법론
(한국웹에서는 보통 학습/해결 방법을 학습하는 방법 이라고 소개한다.)
일반적인 ML에서는 모델의 결과와, 데이터의 레이블에 기반한 
loss(cost) function을 만들어 최소화 시키는 방법으로 학습하는데
$$
f_{loss}(y, \hat{y})
$$
Meta leaning 에서는 보통 문제를 해결하기 위한 다른 함수(e.g. 유사도)를 정의하고
해당 **함수의 기능을 최대화**하는 방식으로 학습한 후
해당 함수를 이용하며 문제를 해결한다.

![](2021-11-14-17-50-48.png)

---
# One-shot learning: refs
- refs
    - [meta learning 영문 wiki](https://en.wikipedia.org/wiki/Meta_learning_(computer_science))
    - [meta learning 영문 blog](https://machinelearningmastery.com/meta-learning-in-machine-learning/)
    - [Supervised vs one-shot 영문](https://searchenterpriseai.techtarget.com/feature/Comparing-semi-supervised-machine-learning-vs-one-shot-learning)
    - [few-shot 영문](https://www.borealisai.com/en/blog/tutorial-2-few-shot-learning-and-meta-learning-i/)
    - [few-shot 한글(카카오)](https://www.kakaobrain.com/blog/106)

---
# Triplets loss: intro
![](2021-11-14-17-57-32.png)

---
# Triplets loss: code
```python
def batch_all_triplet_loss(labels, embeddings, margin, squared=False):
    """Build the triplet loss over a batch of embeddings.

    We generate all the valid triplets and average the loss over the positive ones.

    Args:
        labels: labels of the batch, of size (batch_size,)
        embeddings: tensor of shape (batch_size, embed_dim)
        margin: margin for triplet loss
        squared: Boolean. If true, output is the pairwise squared euclidean distance matrix.
                 If false, output is the pairwise euclidean distance matrix.

    Returns:
        triplet_loss: scalar tensor containing the triplet loss
    """
    # Get the pairwise distance matrix
    pairwise_dist = _pairwise_distances(embeddings, squared=squared)

    anchor_positive_dist = tf.expand_dims(pairwise_dist, 2)
    anchor_negative_dist = tf.expand_dims(pairwise_dist, 1)

    # Compute a 3D tensor of size (batch_size, batch_size, batch_size)
    # triplet_loss[i, j, k] will contain the triplet loss of anchor=i, positive=j, negative=k
    # Uses broadcasting where the 1st argument has shape (batch_size, batch_size, 1)
    # and the 2nd (batch_size, 1, batch_size)
    triplet_loss = anchor_positive_dist - anchor_negative_dist + margin
```
- ref: https://omoindrot.github.io/triplet-loss

---
# Face Verification in the lecture
![](2021-11-14-18-20-36.png)

---
# Face Verification in the lecture
![](2021-11-14-18-20-20.png)

---
# Neural Style Transfer
- Visualize CNN
- Neural Style Transfer
---
# Visualize CNN

---

# Visualize CNN: layer 1
각 layer에서 unit의 최대로 activate 되게 하는 fetch를 보인 것 입니다.
- fetch는 dataset에서 찾습니다.
![](2021-11-14-22-23-18.png)

---
# Visualize CNN: layer 2
![](2021-11-14-22-23-29.png)

---
# Visualize CNN: layer 3
![](2021-11-14-22-37-57.png)

---
# Visualize CNN: layer 4
![](2021-11-14-22-38-30.png)

---
# Visualize CNN: layer 5
![](2021-11-14-22-38-39.png)

---
# Visualize CNN: other method
- Filters의 weight를 그대로 출력한다. 
- Image를 the filters 에 그대로 적용하여 출력한다. 
![](2021-11-14-22-39-34.png)
- https://www.youtube.com/watch?v=ho6JXE3EbZ8

---
# Neural Style Transfer: cost function
![](2021-11-14-22-44-31.png)
![](2021-11-14-22-43-57.png)

---
# Neural Style Transfer: generate
![](2021-11-14-22-42-53.png)
$$
J(S, C, G)
$$
에서 S,C 를 고정하고, gradient descent로 G를 업데이트

---
# Neural Style Transfer: content cost function
 ![](2021-11-14-22-43-14.png) | ![](2021-11-14-22-50-40.png)
-- | --
 | 

---
# Neural Style Transfer: style cost function

![](2021-11-14-23-33-46.png)| ![](2021-11-14-23-29-29.png)
-- | --
 | 

---

### Activation들의 조합/곱은(k, k`) 
- 그림의 스트로크와 같은  
특징(색상, 표현, 패턴 또는 이들의 조합)을 대표 할 것이다.
- Style cost는 generated image와 style image 간에 이 특징들의 차이를 나타낸다.
- Style cost를 최소화 하는 것은   
igenerated image와 style image 간의 특징 차이를 최소화 하는 것!

![](2021-11-14-23-31-00.png) 

 ---

# 1D and 3D Generalizations
## Conv 1d
![](2021-11-14-23-51-34.png)
- ref https://arxiv.org/abs/1408.5882

---
## Conv 3d
![](2021-11-15-00-07-42.png)

---
## Conv 3d: rels
- [Learning Spatiotemporal Features with 3D Convolutional Networks](https://arxiv.org/abs/1412.0767)
- https://github.com/karolzak/conv3d-video-action-recognition

- [conv3d for image classification](https://keras.io/examples/vision/3D_image_classification/)

**Comment by kingheadcat** [comment]
[face_rec.pdf](https://github.com/KimDahye/deeplearning_study/files/7534017/face_rec.pdf)
c.pdf

**Comment by kingheadcat** [comment]
클린봇 데뷰 발표자료
https://deview.kr/data/deview/session/attach/1600_T2_%EC%B5%9C%EA%B2%BD%ED%98%B8_%EC%A1%B0%EA%B8%88%20%EB%8D%94%20%EC%95%84%EB%A6%84%EB%8B%A4%EC%9A%B4%20%EB%8C%93%EA%B8%80%20%EA%B2%BD%ED%97%98%EC%9D%84%20%EC%9C%84%ED%95%B4%EC%84%9C%20Task%20%EC%A0%95%EC%9D%98%20%EB%B6%80%ED%84%B0%20%EC%84%9C%EB%B9%84%EC%8A%A4%EA%B9%8C%EC%A7%80%20Cleanbot%202%20%EA%B0%9C%EB%B0%9C%EA%B8%B0.pdf
